### ISSUE TITLE
About the Implementation of GRPO

### ISSUE THREAD

**Issue by macheng6** [huggingface/trl#2681]
Is the result of x-x.detach() all zero?

**Comment by qgallouedec** [comment]
So it's more a torch question, but I guess you refer to 

https://github.com/huggingface/trl/blob/4659ad916fe3e05fd81bf1297db7a6767d2c69bc/trl/trainer/grpo_trainer.py#L412-L413

Yes, the result is zero, but you can't discard it, you need the gradient, see https://github.com/huggingface/trl/pull/2565#issuecomment-2595837761

**Comment by qgallouedec** [comment]
To close the loop on your question: `per_token_logps - per_token_logps.detach()` does evaluate to all zeros, so `torch.exp(...)` is 1 in the forward pass. But its gradient w.r.t. `per_token_logps` is the identity. That's exactly the point: the loss value stays equal to the advantage while the policy gradient still flows through the log-probs.
